pkg/controller/installation/typha_autoscaler_test.go

```diff
@@ -16,59 +16,119 @@ package installation
-	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	kfake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
 )
 
+// Mock a cache.ListWatcher for nodes to use in the test as there is no other suitable
+// mock available in the fake packages.
+// Ref: https://github.com/kubernetes/client-go/issues/352#issuecomment-614740790
+type nodeListWatch struct {
+	kubernetes.Interface
+}
+
+func (n nodeListWatch) List(options metav1.ListOptions) (runtime.Object, error) {
+	return n.Interface.CoreV1().Nodes().List(context.Background(), options)
+}
+
+func (n nodeListWatch) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	return n.Interface.CoreV1().Nodes().Watch(context.Background(), options)
+}
+
+// Mock a cache.ListWatcher for nodes to use in the test as there is no other suitable
+// mock available in the fake packages.
+// Ref: https://github.com/kubernetes/client-go/issues/352#issuecomment-614740790
+type typhaListWatch struct {
+	kubernetes.Interface
+}
+
+func (t typhaListWatch) List(options metav1.ListOptions) (runtime.Object, error) {
+	return t.Interface.AppsV1().Deployments("calico-system").List(context.Background(), options)
+}
+
+func (t typhaListWatch) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	return t.Interface.AppsV1().Deployments("calico-system").Watch(context.Background(), options)
+}
+
 var _ = Describe("Test typha autoscaler ", func() {
-	var c client.Client
 	var statusManager *status.MockStatus
-
+	var c *kfake.Clientset
 	var ctx context.Context
+	var nlw, tlw cache.ListerWatcher
 
 	BeforeEach(func() {
-		c = fake.NewFakeClientWithScheme(scheme.Scheme)
-		err := c.Create(ctx, &corev1.Namespace{
-			TypeMeta: metav1.TypeMeta{Kind: "Namespace", APIVersion: "v1"},
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "calico-system",
-			},
-		})
-		Expect(err).NotTo(HaveOccurred())
 		statusManager = new(status.MockStatus)
 
+		objs := []runtime.Object{
+			&corev1.Namespace{
+				TypeMeta: metav1.TypeMeta{Kind: "Namespace", APIVersion: "v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "calico-system",
+				},
+			},
+		}
+		c = kfake.NewSimpleClientset(objs...)
+		nlw = nodeListWatch{c}
+		tlw = typhaListWatch{c}
 		ctx = context.Background()
 	})
 
+	It("should initialize an autoscaler", func() {
+		ta := newTyphaAutoscaler(c, nlw, tlw, statusManager)
+		ta.start()
+	})
+
 	It("should get the correct number of nodes", func() {
 		n1 := createNode(c, "node1", map[string]string{"kubernetes.io/os": "linux"})
 		_ = createNode(c, "node2", map[string]string{"kubernetes.io/os": "linux"})
 
-		ta := newTyphaAutoscaler(c, statusManager)
-		schedulableNodes, linuxNodes, err := ta.getNodeCounts()
-		Expect(err).To(BeNil())
-		Expect(schedulableNodes).To(Equal(2))
-		Expect(linuxNodes).To(Equal(2))
+		ta := newTyphaAutoscaler(c, nlw, tlw, statusManager)
```

Comment:
not really, since we are pre-loading the data before calling `ta.start()`, although maybe this does apply to the second update. Either way, I didn't see this test taking very long. I'll look at it.